[#52147] quick action table headers

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/52147 

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Clicking on the table header in the project list offers a variety of actions to perform on this column.

Sort by a column, move columns around, use a filter by a column value, add or remove columns. All this now at the convenience of a click.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

https://github.com/user-attachments/assets/b2afbe5d-377a-4378-8577-7be15ca04250


# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

I went with Primers `ActionMenu` as it seemed like a good fit. Where possible, existing functionality of the project list was kept. For example, remembering the query parameters for previous changes when making further changes (first sort, then change available columns, then sort again).

To interact with the filter component and configure view components, I used Stimulus. All other operations are triggered by visiting a URL.

In general, the quick action table headers should be easily portable to other tables. This is why all of the action menu methods went into the general sort_helper. Model-specific infos such as custom filter mappings (see next paragraph) are provided from the view when calling the component.

About column filter mappings: Some actions are not suitable for certain columns. This depends on the column itself and its position within the table. You cannot move the leftmost column further to the left. At the same time, you cannot filter by columns where we have no fitting filter for. For mapping existing filters to certain columns, I have added constant to `projects_helper.rb`. When applying the action menu to other tables, the models must provide their column-filter-mapping in a similar fashion.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#52147] i18n, rubocop styling

WIP - remove column

WIP - remove column refactor

WIP - move column left or right

WIP - small refactor

WIP - filter by

WIP - style adjustments as discussed with the frontend

EinLama committed:
[#52147] tidy up

EinLama committed:
[#52147] use helpers for action menu construction

Better have many smaller methods than one giant one.

EinLama committed:
[#52147] reorganize quick action table header methods

EinLama committed:
[#52147] further simplifications

EinLama committed:
[#52147] use symbols for icons, doc comment

EinLama committed:
[#52147] remove unnecessary helper method

EinLama committed:
[#52147] fix cop errors

EinLama committed:
[#52147] repair sort helper specs

EinLama committed:
[#52147] helper specs, only move column if possible

We only offer to move a column to the left when it's not the leftmost
column already. Same for the right hand side.

EinLama committed:
[#52147] reduce complexity of column moving

EinLama committed:
[#52147] make feature specs green

EinLama committed:
[#52147] add feature specs for action menu

ulferts commented:
I didn't do a proper review but rather only tried out the PR for a bit. I really enjoyed playing around with it cause the UX felt nicely direct. But while playing around, I stumbled over two things:

<img width="770" alt="image" src="https://github.com/user-attachments/assets/82b04cdc-baa5-4838-b0c1-1fac363c9ec5">

* The table headers are no longer ensured to be fully visible. E.g. "Version" is cut off. That used to be different before.
* There is no quick action on text columns. I get that they are not sortable but one could move or remove such a column via the quick actions.

EinLama committed:
[#52147] ensure columns are always readable

EinLama committed:
[#52147] show the quick action menu for non-sortable columns

Since you can do more things, such as (re)moving a column, it would be a
shame to not offer these options.

For non-sortable columns, we will just not offer the sort-action.

EinLama commented:
Thanks for the feedback, Jens. I agree and have added a style fix as well as the option to use the action menu on non-sortable columns, too.

What's left now is to also allow the "mark as favorite" column to work with the action menu with the same reasoning. Currently looking into that.

EinLama committed:
[#52147] provide sort helper spec

EinLama committed:
[#52147] favorite column provides an action menu

EinLama committed:
[#52147] provide proper styling for icon-only columns

EinLama committed:
[#52147] restore turbo-stream functionality

EinLama committed:
[#52147] comments

EinLama committed:
[#52147] greenify sort_helper specs again

EinLama committed:
[#52147] not all project tables use quick actions

In the project storage module, there is a project list that should not
use these quick actions as the columns there are not sortable.

Same for custom field settings.

EinLama committed:
Merge branch 'dev' into feature/52147-quick-action-table-headers

aaron-contreras commented on app/helpers/sort_helper.rb:
Excellent!

aaron-contreras commented on frontend/src/global_styles/primer/_overrides.sass:
```diff
@@ -117,3 +117,9 @@ sub-header,
 // Todo: Remove once https://github.com/primer/view_components/pull/3087 is merged
 .FormControl-spacingWrapper
   row-gap: var(--stack-gap-normal)
+
+// When our generic table uses the quick action menu, column headers are cut off and no longer fully readable.
+// This fix ensures that column headers are always displayed. An example for such a table is the project list.
+.generic-table
+  .generic-table--action-menu-button
+    min-width: max-content !important
```
👍

aaron-contreras commented on frontend/src/stimulus/controllers/dynamic/filter/filters-form.controller.ts:
```diff
@@ -142,6 +142,11 @@ export default class FiltersFormController extends Controller {
 
+  showDisplayFilters() {
+    this.displayFiltersValue = true;
+    this.displayFiltersValueChanged();
```
Is this actually necessary? IIRC when these are declared Stimulus controller values, they are automatically observed and call the `valueNameValueChanged` without having to invoke it.

EinLama replied:
I know what you mean. I had the same assumption initially and was surprised by this behavior as well.

To be sure, I tried commenting out the call to `showDisplayFilters` again -> the filter component is not toggled automatically when this line is missing.

aaron-contreras reviewed (approved):
🌴 Great work. Left a question as I recall some of the behavior in a Stimulus controller might be duplicated. If not, feel free to merge with this approval

EinLama commented:
Thank you 🎉
